### PR TITLE
Enhanced Balance Display, User Interface Refinements, and Code Optimization

### PR DESCRIPTION
**Pull request by 71walceli**
This pull request focuses on enhancing the user interface (UI) elements related to balance display and incorporates several improvements:

* **Balance Display:**
    * Introduced `BigNumber.js` to handle large numbers and ensure accurate balance representation.
    * Refactored code to include chain decimals and symbols, enabling accurate display of balances with their corresponding units.
    * Implemented an amount formatter for improved readability of displayed values.
    * Added functionalities to display names for amounts, providing a more user-friendly experience.
* **UI Refinements:**
    * Incorporated prev/next buttons to facilitate navigation within identity verification steps.
    * Improved code organization and removed unnecessary code, resulting in better maintainability and performance.
* **Other Improvements:**
    * Added functionalities for handling process state and navigating through verification stages (if applicable).
    * Fixed potential bugs related to account selection and balance retrieval.
* **Events handling:**
    * Frontend now regularly pulls parachain events related to identity and judgment, in order to keep track of extrinsic execution.
    * Verification process now updated according to parachain events, such as _Reasonable_ and _Known Good_ judgments.

These enhancements contribute to a more informative, user-friendly, and efficient application, providing users with a clearer understanding of their balances and streamlining interactions within the UI.

### CODE REVIEW

**Comment by 71walceli** [comment]
**UPDATE**: Initial Implementation of Challenges API and WebSocket Connectivity

This pull request introduces the initial integration of the Challenges API and establishes WebSocket connectivity for real-time updates.

**Key Improvements:**

* **WebSocket Connectivity:** 
  * Implemented a robust WebSocket connection to the Challenges API, ensuring reliable communication.
  * Addressed issues related to unexpected WebSocket closures and reconnections.
* **API Integration:**
  * Loaded the Challenges API URL from environment vars, enabling data retrieval from the API endpoint.
  * Corrected the response data fetching logic to align with the dummy API's structure.
  * Refactored messaging logic to trigger updates based on account address changes, optimizing efficiency.
* **Component Structure:**
  * Renamed components for better clarity and organization.
  * Introduced a custom hook to handle WebSocket API challenges, encapsulating all related code to consume it.
  * Integrated the hook into the `App` component, paving the way for the retrieval and display of challenge data.

This initial implementation lays the groundwork for further development and integration of the Challenges API, providing a foundation for real-time updates and interactive features within the application.

**Comment by 71walceli** [comment]
UPDATE Enhanced Challenge Verification and User Interface Improvements

This pull request introduces significant enhancements to the challenge verification process and user interface:

**Key Improvements:**

* **Challenge Verification:**
    * Integrated WebSocket callbacks to handle real-time updates from the Challenges API.
    * Implemented logic to update the verification state based on notifications received from the API.
    * Added functionality to request new challenges and initiate the verification process.
    * Refactored the WebSocket logic to be specific to the challenge verification screen, improving code organization and maintainability.
* **User Interface:**
    * Enhanced the display of challenges for verification, providing a clear and informative user interface.
    * Removed redundant WebSocket code and unnecessary code elements, optimizing the codebase.
    * Refactored code to improve readability and prevent horizontal scrolling in the IDE.
    * Added a "New Challenge" button to initiate the challenge request process.
    * Updated type information and fixed potential type-related issues.

These improvements contribute to a more robust, user-friendly, and efficient challenge verification process, providing a seamless experience for users.

File: src/components/IdentityVerificationProcess.tsx [hunk 1]
```diff
@@ -1,25 +1,50 @@
...
 import { appState } from '~/App';
-import { SignerProvider } from '@reactive-dot/react';
+import { getSs58AddressInfo } from 'polkadot-api';
+import { IdentityVerificationStates } from '~/constants';
 
+enum Stages {
+  SetIdentityForm = 0,
+  Challenges = 1,
+  Congrats = 2,
+}
```
Comment (hitchho) [inline src/components/IdentityVerificationProcess.tsx]: improve state namings for stages
```
enum RegistrationStages {
  Submission = 0,
  Verification = 1,
  Complete = 2,  
```

File: src/components/IdentityVerificationProcess.tsx [hunk 2]
```diff
@@ -39,54 +64,75 @@ const IdentityVerificationProcess = () => {
...
+    value: appStateSnap.verificationProgress,
+    key: IdentityVerificationStates[appStateSnap.verificationProgress],
+  }), [percentage])
+
+  const canProceed = useMemo((() => 
+    (stage <= Stages.SetIdentityForm 
+      && appStateSnap.verificationProgress >= IdentityVerificationStates.JudgementRequested
+    )
+    || (stage <= Stages.Challenges 
+      && appStateSnap.verificationProgress >= IdentityVerificationStates.IdentityVerifid
```
Comment (hitchho) [inline src/components/IdentityVerificationProcess.tsx]: IdentityVerifid 

typo

File: src/constants.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+export const CHAIN_UPDATE_INTERVAL = 6000
+
+export enum IdentityVerificationStates {
+  Unknown = -1,
+  NoIdentity = 0,
+  IdentitySet = 1,
+  JudgementRequested = 2,
+  FeePaid = 3, // Ready to complete challenges
+  IdentityVerifid = 4, // Judgement is deemed Reasonable, KnownGood, or any state that denotes approval
```
Comment (hitchho) [inline src/constants.ts]: IdentityVerified*

File: src/App.tsx [hunk 4]
```diff
@@ -36,66 +42,348 @@ const DomTitle: React.FC<Props> = ({ route }) => {
...
+      appState.verificationProgress = IdentityVerificationStates.JudgementRequested;
+
+      if (judgementData.find(j => j.state === IdentityJudgement.FeePaid().type)) {
+        appState.verificationProgress = IdentityVerificationStates.FeePaid;
+      }
+      if (judgementData.find(j => [
+        IdentityJudgement.Reasonable().type,
+        IdentityJudgement.KnownGood().type,
+      ].includes(j.state))) {
+        appState.verificationProgress = IdentityVerificationStates.IdentityVerifid;
```
Comment (hitchho) [inline src/App.tsx]: IdentityVerified*

Comment (hitchho) [reply src/App.tsx]: rg -l IdentityVerifid | xargs sed -i "s/IdentityVerifid/IdentityVerified/g"

**Comment by hitchho** [comment]
### 1. Race Conditions in Chain Event Handling 
Current implementation has potential race conditions in event processing:
```typescript
useEffect(() => {
  if (!appStateSnapshot.chain.id || !appStateSnapshot.account?.address) {
    return
  }
  const timer = setInterval(() => {
    handleChainEvent({ type: { pallet, call }, id, onEvent, onError, })
  }, CHAIN_UPDATE_INTERVAL)
  return () => clearInterval(timer)
}, [appStateSnapshot.chain.id, appStateSnapshot.account?.address,])
```

The interval could fire while a previous event is still processing, leading to state inconsistency. Should track pending operations:

```typescript
useEffect(() => {
  if (!appStateSnapshot.chain.id || !appStateSnapshot.account?.address) {
    return;
  }

  let isPending = false;

  const timer = setInterval(async () => {
    // Skip if previous operation still running
    if (isPending) {
      import.meta.env.DEV && console.log('Skipping event check - previous still pending');
      return;
    }

    try {
      isPending = true;
      await handleChainEvent({ 
        type: { pallet, call }, 
        id, 
        onEvent, 
        onError,
      });
    } finally {
      isPending = false;
    }
  }, CHAIN_UPDATE_INTERVAL);

  return () => clearInterval(timer);
}, [appStateSnapshot.chain.id, appStateSnapshot.account?.address]);
```

### 2. Memory Leak Risk
Current implementation has unbounded memory growth:
```typescript
const relevantBlocks = useRef([])
```
This array grows indefinitely as new blocks are processed. Need to implement block pruning mechanism.

### 3. Unprotected State Transitions
Chain selection lacks validation:
```typescript
const handleChainSelect = (key: string) => {
  startTransition(() => {
    appState.chain = {
      id: key,
      ss58Format: null,
      tokenDecimals: null, 
      tokenSymbol: null,
    };
    setCustomSelected(false);
    setIsOpen(false);
  });
};
```
Nothing prevents calling this with invalid chain IDs that could crash the app. Should validate key against config.chains.

## Architectural Issues

### Circular Dependencies
There are circular dependencies between IdentityForm.tsx and App.tsx. Both files import:
```typescript
import { useTypedApi } from '@reactive-dot/react';
import { useSnapshot } from 'valtio';
import { appState } from '~/App';
import { CHAIN_UPDATE_INTERVAL, IdentityVerificationStates } from '~/constants';
import { useIdentityEncoder } from '~/hooks/hashers/identity';
```

Additionally, IdentityForm is importing `appState` from `App.tsx` while `App.tsx` has all the state definitions.

## Recommended Solution

1. Move all state management to dedicated Valtio stores:
```typescript
// src/store/index.ts
export * from './identityStore'
export * from './chainStore'
export * from './accountStore'
export * from './challengeStore'
```

2. Split monolithic appState into focused stores with proper typings and encapsulated actions:
```typescript
// src/store/identityStore.ts
interface IdentityState {
  data?: Identity
  verificationProgress: IdentityVerificationStates
  // ...
}

export const identityStore = proxy<IdentityState>({
  // State and actions
})
```

This approach will:
- Eliminate circular dependencies
- Provide better type safety
- Enable focused state updates
- Make testing easier
- Improve code maintainability




fg.

```
// src/store/types.ts
export interface ActionResult<T> {
  success: boolean
  data?: T
  error?: string
}

export enum IdentityVerificationStates {
  Unknown = -1,
  NoIdentity = 0,
  IdentitySet = 1,
  JudgementRequested = 2,
  FeePaid = 3,
  IdentityVerified = 4,
}

export interface Identity {
  displayName: string
  matrix: string
  email: string
  discord: string
  twitter: string
}

export interface ChainInfo {
  id: string
  ss58Format?: number
  tokenDecimals?: number
  tokenSymbol?: string
}

export interface Block {
  hash: string
  number: number
  timestamp: number
  data: any // Define specific block data structure
}

export interface Challenge {
  value: string
  verified: boolean
}

// src/store/accountStore.ts
import { proxy } from 'valtio'
import { subscribeKey } from 'valtio/utils'
import { LocalStorageUtil } from '~/utils'

interface AccountState {
  address?: string
  name?: string
  balance?: {
    free: bigint
    reserved: bigint
    frozen: bigint
    flags: bigint
  }
  actions: {
    setAccount: (address: string, name: string) => Promise<ActionResult<void>>
    updateBalance: (balance: AccountState['balance']) => void
    disconnect: () => void
  }
}

export const accountStore = proxy<AccountState>({
  actions: {
    setAccount: async (address: string, name: string) => {
      try {
        accountStore.address = address
        accountStore.name = name
        LocalStorageUtil.setItem('account', { address, name })
        return { success: true }
      } catch (error) {
        return { success: false, error: error.message }
      }
    },
    updateBalance: (balance) => {
      accountStore.balance = balance
    },
    disconnect: () => {
      accountStore.address = undefined
      accountStore.name = undefined
      accountStore.balance = undefined
      LocalStorageUtil.removeItem('account')
    }
  }
})

// src/store/chainStore.ts
import { proxy } from 'valtio'
import { config } from '~/api/config'

interface ChainState {
  current: ChainInfo
  relevantBlocks: Block[]
  maxBlockHistory: number
  actions: {
    setChain: (chainId: string) => Promise<ActionResult<void>>
    addBlock: (block: Block) => void
    pruneBlocks: () => void
    updateChainMeta: (meta: Partial<ChainInfo>) => void
  }
}

export const chainStore = proxy<ChainState>({
  current: {
    id: import.meta.env.VITE_APP_DEFAULT_CHAIN || Object.keys(config.chains)[0]
  },
  relevantBlocks: [],
  maxBlockHistory: 100,
  actions: {
    setChain: async (chainId) => {
      try {
        if (!config.chains[chainId]) {
          throw new Error(`Invalid chain ID: ${chainId}`)
        }
        chainStore.current = { id: chainId }
        return { success: true }
      } catch (error) {
        return { success: false, error: error.message }
      }
    },
    addBlock: (block) => {
      chainStore.relevantBlocks.push(block)
      if (chainStore.relevantBlocks.length > chainStore.maxBlockHistory) {
        chainStore.actions.pruneBlocks()
      }
    },
    pruneBlocks: () => {
      const excessBlocks = chainStore.relevantBlocks.length - chainStore.maxBlockHistory
      if (excessBlocks > 0) {
        chainStore.relevantBlocks.splice(0, excessBlocks)
      }
    },
    updateChainMeta: (meta) => {
      chainStore.current = { ...chainStore.current, ...meta }
    }
  }
})

// src/store/identityStore.ts
import { proxy } from 'valtio'
import { subscribeKey } from 'valtio/utils'

interface IdentityState {
  data?: Identity
  verificationProgress: IdentityVerificationStates
  fees: {
    requestJudgement?: bigint
    setIdentityAndRequestJudgement?: bigint
  }
  hashes: {
    identityOf?: Uint16Array
  }
  actions: {
    update: (identity: Partial<Identity>) => Promise<ActionResult<void>>
    setProgress: (state: IdentityVerificationStates) => void
    updateFees: (fees: Partial<IdentityState['fees']>) => void
    updateHash: (hash: Uint16Array) => void
    reset: () => void
  }
}

export const identityStore = proxy<IdentityState>({
  verificationProgress: IdentityVerificationStates.Unknown,
  fees: {},
  hashes: {},
  actions: {
    update: async (newIdentity) => {
      try {
        identityStore.data = { ...identityStore.data, ...newIdentity }
        return { success: true }
      } catch (error) {
        return { success: false, error: error.message }
      }
    },
    setProgress: (state) => {
      identityStore.verificationProgress = state
    },
    updateFees: (fees) => {
      identityStore.fees = { ...identityStore.fees, ...fees }
    },
    updateHash: (hash) => {
      identityStore.hashes.identityOf = hash
    },
    reset: () => {
      identityStore.data = undefined
      identityStore.verificationProgress = IdentityVerificationStates.Unknown
      identityStore.fees = {}
      identityStore.hashes = {}
    }
  }
})

// Persist identity changes
subscribeKey(identityStore, 'data', (identity) => {
  if (identity) {
    LocalStorageUtil.setItem('identity', identity)
  }
})

// src/store/challengeStore.ts
interface ChallengeState {
  challenges: Record<string, Challenge>
  actions: {
    setChallengeValue: (field: string, value: string) => void
    setVerified: (field: string, verified: boolean) => void
    reset: () => void
  }
}

export const challengeStore = proxy<ChallengeState>({
  challenges: {},
  actions: {
    setChallengeValue: (field, value) => {
      challengeStore.challenges[field] = {
        ...challengeStore.challenges[field],
        value
      }
    },
    setVerified: (field, verified) => {
      challengeStore.challenges[field] = {
        ...challengeStore.challenges[field],
        verified
      }
    },
    reset: () => {
      challengeStore.challenges = {}
    }
  }
})

// src/store/wsStore.ts
interface WsState {
  connected: boolean
  error: string | null
  isPending: boolean
  actions: {
    connect: (url: string) => Promise<ActionResult<void>>
    disconnect: () => void
    clearError: () => void
  }
}

export const wsStore = proxy<WsState>({
  connected: false,
  error: null,
  isPending: false,
  actions: {
    connect: async (url) => {
      try {
        if (wsStore.isPending) {
          return { success: false, error: 'Connection attempt in progress' }
        }
        wsStore.isPending = true
        
        // WebSocket connection logic here
        // Update other stores on messages
        wsStore.connected = true
        wsStore.error = null
        return { success: true }
      } catch (error) {
        wsStore.error = error.message
        return { success: false, error: error.message }
      } finally {
        wsStore.isPending = false
      }
    },
    disconnect: () => {
      wsStore.connected = false
      wsStore.error = null
      // Cleanup logic
    },
    clearError: () => {
      wsStore.error = null
    }
  }
})

// src/store/index.ts
export * from './types'
export * from './accountStore'
export * from './chainStore'
export * from './identityStore'
export * from './challengeStore'
export * from './wsStore'

// Store initialization
export const initializeStores = async () => {
  // Load persisted data
  const savedIdentity = LocalStorageUtil.getItem<Identity>('identity')
  if (savedIdentity) {
    await identityStore.actions.update(savedIdentity)
  }

  const savedAccount = LocalStorageUtil.getItem<{ address: string; name: string }>('account')
  if (savedAccount) {
    await accountStore.actions.setAccount(savedAccount.address, savedAccount.name)
  }

  // Initialize WebSocket connection
  if (import.meta.env.VITE_APP_WS_URL) {
    await wsStore.actions.connect(import.meta.env.VITE_APP_WS_URL)
  }
}

// Reset all stores
export const resetStores = () => {
  identityStore.actions.reset()
  challengeStore.actions.reset()
  accountStore.actions.disconnect()
  wsStore.actions.disconnect()
}

// Example usage in a component:
/*
import { useSnapshot } from 'valtio'
import { identityStore, chainStore, initializeStores } from '~/store'

const MyComponent = () => {
  const { data: identity, verificationProgress } = useSnapshot(identityStore)
  const { current: chain } = useSnapshot(chainStore)

  useEffect(() => {
    initializeStores()
  }, [])

  const handleUpdateIdentity = async () => {
    const result = await identityStore.actions.update({
      displayName: 'New Name'
    })
    if (!result.success) {
      console.error(result.error)
    }
  }

  return (
    // Component JSX
  )
}
*/
```

File: src/components/IdentityVerificationProcess.tsx (hunk 1 above)
Comment (71walceli) [reply src/components/IdentityVerificationProcess.tsx]: Commit 5f4b490f251205a34b8049f805dc8bf6dc16d652

File: src/components/IdentityVerificationProcess.tsx (hunk 2 above)
Comment (71walceli) [reply src/components/IdentityVerificationProcess.tsx]: Corrected in commit 111ecd16674c163a521f73d7c0af0f6450a449d7

File: src/constants.ts (hunk 3 above)
Comment (71walceli) [reply src/constants.ts]: 111ecd16674c163a521f73d7c0af0f6450a449d7

**Comment by 71walceli** [comment]
> ### 1. Race Conditions in Chain Event Handling
> Current implementation has potential race conditions in event processing:
> 
> ```ts
> useEffect(() => {
>   if (!appStateSnapshot.chain.id || !appStateSnapshot.account?.address) {
>     return
>   }
>   const timer = setInterval(() => {
>     handleChainEvent({ type: { pallet, call }, id, onEvent, onError, })
>   }, CHAIN_UPDATE_INTERVAL)
>   return () => clearInterval(timer)
> }, [appStateSnapshot.chain.id, appStateSnapshot.account?.address,])
> ```
> 
> The interval could fire while a previous event is still processing, leading to state inconsistency. Should track pending operations:
> 
> ```ts
> useEffect(() => {
>   if (!appStateSnapshot.chain.id || !appStateSnapshot.account?.address) {
>     return;
>   }
> 
>   let isPending = false;
> 
>   const timer = setInterval(async () => {
>     // Skip if previous operation still running
>     if (isPending) {
>       import.meta.env.DEV && console.log('Skipping event check - previous still pending');
>       return;
>     }
> 
>     try {
>       isPending = true;
>       await handleChainEvent({ 
>         type: { pallet, call }, 
>         id, 
>         onEvent, 
>         onError,
>       });
>     } finally {
>       isPending = false;
>     }
>   }, CHAIN_UPDATE_INTERVAL);
> 
>   return () => clearInterval(timer);
> }, [appStateSnapshot.chain.id, appStateSnapshot.account?.address]);
> ```

Well, that was refactored into following code:
```ts
const getEffectCallback = ({type: { pallet, call }, onEvent, onError, id}) => {
  return () => {
    if (!appStateSnapshot.chain.id || !appStateSnapshot.account?.address) {
      return
      }
      const timer = window.setInterval(() => {
        handleChainEvent({ type: { pallet, call }, id, onEvent, onError, })
      }, CHAIN_UPDATE_INTERVAL)
      return () => window.clearInterval(timer)
    }
  }
```

None of the events that are needed, such as `Identity.IdentitySet`, `Identity.IdentityCleared`, `Identity.JrdgementRequested`, `Identity.JrdgementGiven` are meant to affect same state. Also, events are filtered by current user's account, so they may not really interfere with one another. 

Only race condition that might reasonably happen is to get `Identity.JrdgementRequested` before `Identity.IdentitySet`.

**Comment by 71walceli** [comment]
Ths coe on `App.tsx` is responsible for all of it

```ts
  useEffect(getEffectCallback({ type: { pallet: "Identity", call: "IdentitySet" }, id: "idSet",
    onEvent: data => {
      appState.verificationProgress = 
        // As we do batch calls, we need to know if judgeent is already awaiting
        appStateSnapshot.verificationProgress === IdentityVerificationStatuses.NoIdentity
          ? IdentityVerificationStatuses.IdentitySet
          : appStateSnapshot.verificationProgress
    },
    onError: error => {},
  }), [appStateSnapshot.chain.id, appStateSnapshot.account?.address,])
  
  // ...
  
  useEffect(getEffectCallback({ type: { pallet: "Identity", call: "JudgementRequested" }, 
    id: "judgRequested",
    onEvent: data => {
      appState.verificationProgress = IdentityVerificationStatuses.JudgementRequested
    },
    onError: error => {},
  }), [appStateSnapshot.chain.id, appStateSnapshot.account?.address,])
```

We do take it into account the case where it might get requested judgement event before one of identity having been set.

**Comment by 71walceli** [comment]
Tho it might not be the most effective way to handle it. Recommended approach is being considered.

**Comment by hitchho** [comment]
While you've addressed the ordering of Identity events by checking verification states, the underlying concern is broader than just `IdentitySet` vs `JudgementRequested` ordering. The current pattern of multiple independent intervals could still lead to race conditions in event processing itself, regardless of state updates.

Consider consolidating these event subscriptions into a single manager that can properly coordinate all identity-related events and their processing order. This would also make the code more maintainable than having multiple separate effect callbacks.

The other issues regarding memory management and architectural structure still need to be addressed. Would you like to discuss specific approaches for those?

**Comment by hitchho** [comment]
The current event handling approach is fundamentally broken. Multiple interval timers querying chain events independently creates a fragile, race-prone design that will fail in unpredictable ways.

Consider:

```typescript
const eventManager = proxy({
  isPending: false,
  
  async processEvent(pallet: string, call: string, data: any) {
    // explicit FSM transitions prevent impossible states
    switch (`${pallet}.${call}`) {
      case 'Identity.IdentitySet':
        return identityStore.progress === NO_IDENTITY ? SET : identityStore.progress
        
      case 'Identity.JudgementRequested': 
        return REQUESTED
        
      default:
        throw new Error(`invalid event: ${pallet}.${call}`)
    }
  },

  // single event source eliminates races
  async poll() {
    if (this.isPending) return
    
    try {
      this.isPending = true
      await Promise.all(EVENTS.map(async ({pallet, call}) => {
        const data = await typedApi.event[pallet][call].pull()
        this.processEvent(pallet, call, data) 
      }))
    } finally {
      this.isPending = false
    }
  }
})

// single interval eliminates coordination complexity
setInterval(eventManager.poll, INTERVAL)
```

The current implementation is grossly overengineered. A single event loop with explicit state machine transitions is both simpler and more correct. Multiple intervals polling the same events is nonsensical.

This also needs proper error handling and cleanup. The current code leaks resources and fails silently.

**Comment by 71walceli** [comment]
ok sir. Thanls for your adivce... I'll consider those ways to fix it..

**Comment by 71walceli** [comment]
About the `relevantBlocks`, ihose are meant to store the blocksextrinsics that have to do with the events described. Yeah, we don't need to store all events from all time. up to 10 or 20, as long as it tales to get finalized blocks. Excuse me on that one...

**Review by hitchho — APPROVED** [review]

**Pull request merged**